### PR TITLE
Template Only Or HBS Strategy 

### PR DESCRIPTION
**Pull request by betocantu93**
# Why
Mainly to allow splattributes and modifiers, maybe when embroider lands we can get tree shaking working for unused svgs!

## TL;DR
For every svg => 1 template only hbs
```hbs
 <Svg @name="svg/some/nested/svg />
 <Svg::Some::Nested::Svg  style="background: red;" {{on "mouseenter" this.something}} />
```

1. Updates dependencies
 - In order to create template only components I decided to upgrade the deps.
2. Hbs strategy
 - For every inline svg found, create an hbs file with this template for now

```hbs
const SVGAsHSBTemplate = ({ attrs, content }) => `
  <svg ${formatAttrs(attrs)} ...attributes>
    {{#if @title}}
      <title id={{@titleId}}> {{@title}} </title>
    {{/if}}
    ${content}
  </svg>
`;
```
By default, it adds a namespace or prefix => "Svg"
It respects the folder structure, so for public/images/my/image/hola.svg ---> `<Svg::Images::My::Image::Hola />`

You can also provide another idGen function to generate files name if needed.

4. Adds `<Svg />` component to ergonomically call stuff like `<Svg @name="svg/images/my/image/hola" />`

5. Updates viewer and copypasta

Screen Shot 2020-10-12 at 14 53 55

There a few restrictions with naming of the assets, since components name rules.

`.` will be removed
`_` replaced by dashes
`/` for `<Svg @name="svg/car/blue" />` invocation, nothing happens, for component invocation, use `::`  `<Svg::Car::Blue />`
`toLowerCase` , so for myBlueCar === mybluecar `<Svg @name="svg/mybluecar" />

# Updates
After Ember Contributors Workshop, @ef4 helped me out to import these generated template only components on demand with the native import, so now they kind of work like normal images, they are not included in the final bundle, instead they are built as separate js files and placed in /public/ember-svg-jar/components/svg by default, and the new invocation would be 

```hbs
<Svg @name="svg/my-icon" />
```

This component will handle the dynamic native import for you and all that internally.

I also updated the copypasta to generate the previously mentioned snipet, we might have some issues regarding fingerprinting, for now, I included ember-cli-resolve-asset, to manage handling fingerprinting

Screen Shot 2020-10-17 at 3 18 22

### CODE REVIEW

**Comment by knownasilya** [comment]
How about a strategy like https://github.com/martndemus/ember-font-awesome/tree/master/lib looks like build time components that get compiled out, could still have modifiers and splattributes

**Comment by betocantu93** [comment]
@knownasilya woah, that's a super nice approach, totally missed it, a few top of the mind questions

- Does that approach support all sort of dynamic edge cases? for example {{fa-icon this.model.icon}}? I think for that one you must use some sort of whitelist 
- I'm not sure how splattribues could work with an AST approach, need to dig more into this to have a real sense of how this could be done statically

//wrapper.hbs
`<FaIcon @name={{@model.name}} ...attributes/>`

`<Wrapper @model={{this.model}} arialabel={{this.model.name}}  />`

Here's a sample of the component I'll be using
[eui-icon](https://github.com/prysmex/ember-eui/blob/master/addon/components/eui-icon/index.ts)

**Comment by GCheung55** [comment]
How will this “shake” out with embroider - will unused SVG files/templates be tree-shaken?

**Comment by betocantu93** [comment]
> How will this “shake” out with embroider - will unused SVG files/templates be tree-shaken?

I'm not entirely sure yet TBH, haven't dig into embroider yet, but as long as the full component invocation is used and staticComponents is enabled perhaps it works automatically 🙏? But if you will use the dynamic name approach I.e. `<Svg @name="svg/some/svg" />` we probably will have to hint embroider with a static whitelist in ember-cli-build 

Honestly the best approach I can think of is if we can chunk every component and let the Svg component dynamically import all of these chunks in runtime using ember-auto-import, but I think ember-auto-import, dynamic imports don't work with dynamic strings, yet...

This is what React svgr does with

https://react-svgr.com/

**Comment by betocantu93** [comment]
Another thing to take into account regarding bundle size and shaking is that the current default inline mode does ship with the full json with all the necessary data to rebuild them in runtime with the helper so in terms of size I think(?) It's kind of the same, but now you can use the power of splattributes, modifiers and accessibility

**Comment by SergeAstapov** [comment]
@betocantu93 this is great! Will this strategy work in addon?

**Comment by betocantu93** [comment]
@SergeAstapov oh yep, that's exactly why I did it, I'm working on this UI lib [ember-eui](https://github.com/prysmex/ember-eui) (its a React port) and I wanted a 1:1 API with their EuiIcon lol, and its really interesting how they handle it here [EuiIcon React version](https://github.com/elastic/eui/blob/01c6986afdd7321f8e86b433be4892345aa7c0fe/src/components/icon/icon.tsx#L629), anyway, AFAIK, ember-svg-jar only needs paths to where to look for svgs, so the host app must declare them, you can see an example in the dummy app of [ember-eui](https://github.com/prysmex/ember-eui) or [ember-polaris](https://github.com/smile-io/ember-polaris)

You just gotta make sure you give the right ergonomics for the users of your addon I guess

File: lib/hbs-packer.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  fs.mkdirSync(dirname);
+}
+
+
+const templateOnlyComponent = (svg) => `
+  import templateOnly from '@ember/component/template-only';
+  import { setComponentTemplate } from '@ember/component';
+  import { hbs } from 'ember-cli-htmlbars';
+
+  export default setComponentTemplate(hbs\`${svg}\`, templateOnly());
```
Comment (SergeAstapov) [inline lib/hbs-packer.js:25]: @betocantu93 does it mean templates won't be pre-compiled at build time and instead compiled at run time once they are loading on client-side?

Comment (betocantu93) [reply lib/hbs-packer.js:25]: Nop, they are compiled at build time but each svg in a separate file, I'm using treeForAddon to accomplish this

**Comment by jherdman** [comment]
To simplify reviewing this, would you be amenable to breaking into two pieces?

1. Upgrade
2. The new feature

**Comment by betocantu93** [comment]
> To simplify reviewing this, would you be amenable to breaking into two pieces?
> 
> 1. Upgrade
> 2. The new feature

Yess, Will do asap @jherdman ! Thanks

**Comment by betocantu93** [comment]
@jherdman  I thiiink I got the PR's! :) it was actually quite a ride to get the tests green

**Comment by betocantu93** [comment]
closed in favor of #185 and #186
